github/app.go

```diff
@@ -0,0 +1,50 @@
+package github
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+)
+
+const tokenDuration = time.Minute * 5
+
+type AppConfig struct {
+	AppID   string
+	KeyPath string
+}
+
+type App struct {
+	jwt       string
+	createdAt time.Time
+}
+
+// NewApp authenticates as a GitHub App.
+//
+// https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app
+func NewApp(config *AppConfig) (*App, error) {
+	signedToken, err := generateJWTToken(&JwtTokenConfig{
+		AppID:      config.AppID,
+		KeyPath:    config.KeyPath,
+		Expiration: tokenDuration,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate JWT token: %w", err)
+	}
+
+	return &App{
+		jwt:       signedToken,
+		createdAt: time.Now(),
+	}, nil
+}
+
+func (a *App) Do(request *http.Request) (*http.Response, error) {
+	if time.Since(a.createdAt) > tokenDuration {
```

Comment:
Forse anzi che fare errore qua e aspettarsi che il chiamate rinnovi il token io metterei la logica per il rinnovo qua no?

Reply:
Rinnovarlo corrisponderebbe a chiamare again `NewApp`, visto che nessun campo si conserva. Lo facciamo qui e modifichiamo tutti i campi?

Reply:
Ah pensavo essendo un bearer ci fosse un modo di refresharlo tipo. Se bisogna rifare il login comunque si io lo farei qua dentro. Per me ci sta chiamare newApp e poi aggiornare i campi. Probabilmente puoi cavartela senza aggiornare a manina i campi e usando invece `copy`.

Reply:
Il fatto è che in teoria questo token è solamente una via per accedere agli altri token, per questo è così short lived. Se vedi infatti in `initializeClient` faccio

```go
func initializeClient() (*github.Client, error) {
	app, err := csunibo.NewApp(&csunibo.AppConfig{
		AppID:   config.AppID,
		KeyPath: config.PrivateKeyPath,
	})
	if err != nil {
		return nil, err
	}

	return app.AuthenticateAsInstallation(config.InstallationID)
}
```

quell'ultimo `AuthenticateAsInstallation` mi genera un altro token che è invece long-lived. Vogliamo comunque refreshare quel token in questo modo?

Reply:
Ah aspetta, avevo capito che il metodo `Do` servisse per fare le richieste vere e proprie all'API (tipo crea PR, commenta, ecc). Quello è opportuno che faccia il refresh in automatico si. Il token intermedio (che immagino funzioni un po' alla oauth) possiamo non mettere la logica per fare il refresh, tanto (se è come oauth) viene usato una sola volta e poi si va avanti con quello a lunga durata.
TL;DR il comportamento che mi aspetterei da `Do` è che gestisca in automatico il refresh del token, qualunque esso sia, mi di errore se fallsice a fare refresh e poi lanci la richiesta, che è quello che deve veramente fare. Penso sia bene mantenere questo livello di astrazione (visto che astrae l'aggiunta del token agli header mi pare giusto astrarre anche il refresh del token).

Reply:
Ti linko la parte della doc che dice a cosa serve quel Do, ovvero ad operare come App: https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app#about-authentication-as-a-github-app

> You must authenticate as a GitHub App in order to make REST API requests as the application. For example, if you want to use the API to generate an installation access token for accessing organization resources, list installations across organizations for your app, or suspend an app installation, you must authenticate as an app.

Pusho anche il refresh automatico

Reply:
Okay, leggendo forse viene fatto meno di quanto pensassi, però comunque ci sta gestire il refresh di questo token se non è troppo complicato. Da qualche parte va fatto, meglio astrarlo secondo me.